### DISCUSSION TITLE
How to manipulate Relay Store

### DISCUSSION THREAD

**Discussion by juliomerisio** [sibelius/relay-workshop-old#3] [Q&A] [answered]
I would like to understand how the Relay store works and how can I access and manipulate data there 

a couple of questions to start:

- can I have multiple stores? If so, how can I identify each one?
- How can I access a connection inside my store? 
- How can I delete a node inside a connection?
- How can I update a node inside a connection?

**Comment by arantespp** [comment, marked as answer]
> can I have multiple stores? If so, how can I identify each one?

Store and environment are linked to each other, so when you said "multiples stores," I can assume that you're talking about multiples environments. However, I don't know if each environment has a unique identifier. So I need to search more about it.

And yes, you can have multiples environments in your app, you just need to wrap the `RelayEnvironmentProvider` with a specific environment.

 > How can I access a connection inside my store?

You can access it in an `updater` callback (see examples below using updater). I've never used outside an updater and I think trying to access it outside `updater` isn't a Relay best practice.

> How can I delete a node inside a connection?

```javascript
updater: (store) => {
  const userProxy = store.get(userId);
  
  if (!userProxy) {
    return;
  }
  
  const projectsConnection = ConnectionHandler.getConnection(
    userProxy,
    'ProjectsListPaginationQuery_ProjectsListPagination_projects',
  );
  
  if (!projectsConnection) {
    return;
  }
  
  ConnectionHandler.deleteNode(projectsConnection, projectId);
},
```

> How can I update a node inside a connection?

```javascript
updater: (store, { makeAppointment: { id } }) => {
  const user = store.get(variables.patientId);
  const appointments = user?.getLinkedRecord('appointments');
  const edges = appointments?.getLinkedRecords('edges');
  const newEdge = store.get(id);
  if (newEdge && edges) {
    const newEdges = [...edges, newEdge];
    // eslint-disable-next-line no-unused-expressions
    appointments?.setLinkedRecords(newEdges, 'edges');
  }
},
```

**Reply by arantespp** [reply]
Note. You can access Store using [`commitLocalUpdate` method](https://relay.dev/docs/guided-tour/updating-data/local-data-updates/).

**Reply by juliomerisio** [reply]
interesting, can we break down the second example? 

1. get the proxy that owns the record
2. get a record inside the proxy
3. set record using the record proxy

**Reply by juliomerisio** [reply]
```js
  const projectsConnection = ConnectionHandler.getConnection(
    userProxy,
    'ProjectsListPaginationQuery_ProjectsListPagination_projects', 
                 // 👆 - How can I find this name?  Is it generated? Is it random? 
                 // Is there a way to have auto-complete assuming that I'm using typescript?
  );
  
```

**Reply by arantespp** [reply]
Sure. My schema:

```graphql
interface Node {
  id: ID!
}

interface Connection {
  nextToken: String
}

type Patient implements UserNode {
  id: ID!
  appointments(
    start: String
    end: String
    limit: Int
    nextToken: String
  ): AppointmentConnection!
}

type AppointmentConnection implements Connection {
  edges: [Appointment!]
  nextToken: String
}

type Appointment implements Node {
  id: ID!
}
```

Updater:

```javascript
updater: (store, { makeAppointment: { id } }) => {
  // Get user proxy, which represents user data handled by Relay.
  const user = store.get(variables.patientId);
  
  // Get appointments of the user.
  const appointments = user?.getLinkedRecord('appointments');
  
  // Get edges of the appoints of the user.
  const edges = appointments?.getLinkedRecords('edges');
  
  // I get the edge that Relay created during mutation. I get this value just passing the node id (we can get any node this way).
  const newEdge = store.get(id);
  
  if (newEdge && edges) {
    // I create the new edges array. You can handle this array using any array helper, like `filter`, `splice`, `sort`...
    const newEdges = [...edges, newEdge];
    
    // Set new edges.
    appointments?.setLinkedRecords(newEdges, 'edges');
  }
}
```

**Reply by arantespp** [reply]
`ProjectsListPaginationQuery_ProjectsListPagination_projects` is a [connection key](https://relay.dev/docs/next/guided-tour/list-data/rendering-connections/).

```javascript
const { data } = usePaginationFragment(
  graphql`
    fragment ProjectsList_ProjectsListPagination on User
    @argumentDefinitions(first: { type: "Int" }, after: { type: "String" })
    @refetchable(queryName: "ProjectsListPagination") {
      projects(first: $first, after: $after)
        @connection(
          key: "ProjectsListPaginationQuery_ProjectsListPagination_projects"
        ) {
        edges {
          node {
            id
            ...ProjectsList_ProjectCard
          }
        }
      }
    }
  `,
  fragmentRef,
);
```

**Reply by juliomerisio** [reply]
The naming convention is kinda tricky for me,  any thoughts around that? Any logical theory behind?

**Reply by arantespp** [reply]
I named because `ProjectsListPaginationQuery_ProjectsListPagination_projects`:

- `ProjectsListPaginationQuery`: it's the query in which I'm paginating. I'm querying on type `User`.
- `ProjectsListPagination`: it's the name of this specific pagination. The same query may have multiples pagination.
- `projects`: it's the name of the property in which I'm paginating. In this case, from `User`, I'm paginating `projects`.

**Reply by juliomerisio** [reply]
such a great explanation @arantespp , thank's for helping me solve my doubts
